<?xml version="1.0"?>
<launch>
  <arg name="serial" default="171340" />
  <arg name="monoserial" default="4103203953" />
  <arg name="camera" default="camera" />
  <arg name="frame_id" default="$(arg camera)_optical_frame" />
  <arg name="rgb_frame_id" default="$(arg camera)_rgb_optical_frame" />

  <!-- Logging -->
  <env name="ROSCONSOLE_CONFIG_FILE" value="$(find ensenso)/config/rosconsole.conf"/>

  <!-- Display hack to avoid 'freeglut (NxLib): failed to open display' when
       logging in via SSH.  Still requires that a window manager is running,
       i.e., a user has to be logged into the graphical environment, for
       example via autologin:

       Create a file /etc/lightdm/lightdm.conf.d/90-autologin.conf with the following content:

       ```
       [Seat:*]
       autologin-user=robot
       autologin-user-timeout=0
       ```
       -->
  <arg name="display" default=":0.0" />
  <env name="DISPLAY" value="$(arg display)" />

  <!-- Ensenso driver -->
  <node name="ensenso_driver" pkg="ensenso" type="ensenso_driver" ns="$(arg camera)" output="screen">
    <!-- ******************** STATIC PARAMETERS ******************** -->
    <param name="serial"                type="string" value="$(arg serial)" />
    <param name="monoserial"                type="string" value="$(arg monoserial)" />
    <param name="camera_frame_id"       type="string" value="$(arg frame_id)" />
    <param name="rgb_camera_frame_id"       type="string" value="$(arg rgb_frame_id)" />
    <param name="stream_calib_pattern"  type="bool"   value="false" />
    <!-- ******************** STREAM PARAMETERS ******************** -->
    <param name="Cloud"  type="bool"   value="true" />
    <param name="Images"  type="bool"   value="true" />
    <param name="Depth"  type="bool"   value="true" />
    <!-- ******************** DYNAMIC PARAMETERS ******************** -->
    <!-- Capture parameters -->
    <param name="AutoBlackLevel"   type="bool"   value="true" />
    <param name="AutoExposure"     type="bool"   value="true" />
    <param name="AutoGain"         type="bool"   value="true" />
    <param name="Binning"          type="int"    value="1" />
    <param name="BlackLevelOffset" type="double" value="0.55" />
    <param name="DisparityMapAOI"  type="bool"   value="false" />
    <param name="Exposure"         type="double" value="1.5" />
    <param name="FlexView"         type="bool"   value="true" />
    <param name="FlexViewImages"   type="int"    value="4" />
    <param name="FrontLight"       type="bool"   value="false" />
    <param name="Gain"             type="double" value="1.0" />
    <param name="GainBoost"        type="bool"   value="false" />
    <param name="HardwareGamma"    type="bool"   value="true" />
    <param name="Hdr"              type="bool"   value="false" />
    <param name="PixelClock"       type="int"    value="24" />
    <param name="Projector"        type="bool"   value="true" />
    <param name="TargetBrightness" type="int"    value="180" />
    <param name="TriggerMode"      type="int"    value="0" />          <!-- Software: 0, FallingEdge: 1, RisingEdge: 2 -->
    <param name="RGBTriggerDelay"      type="int"    value="10" />

    <!-- Stereo parameters -->
    <param name="MinimumDisparity"      type="int"    value="-140" />
    <param name="NumberOfDisparities"   type="int"    value="140" />
    <param name="Scaling"               type="double" value="1.0" />
    <param name="OptimizationProfile"   type="int"    value="2" />     <!-- Aligned: 0, Diagonal: 1, AlignedAndDiagonal: 2 -->
    <param name="DepthChangeCost"       type="int"    value="5" />
    <param name="DepthStepCost"         type="int"    value="30" />
    <param name="ShadowingThreshold"    type="int"    value="1" />

    <!-- Postprocessing parameters -->
    <param name="UniquenessRatio"           type="int" value="10" />
    <param name="MedianFilterRadius"        type="int" value="1" />
    <param name="SpeckleComponentThreshold" type="int" value="3" />
    <param name="SpeckleRegionSize"         type="int" value="150" />
    <param name="FillBorderSpread"          type="int" value="3" />
    <param name="FillRegionSize"            type="int" value="150" />
    <param name="FindPattern"            type="bool" value="true" />
    <!-- Render parameters -->
    <param name="SurfaceConnectivity"            type="int" value="3" />
    <param name="NearPlane"            type="int" value="50" />
    <param name="FarPlane"            type="int" value="3000" />
    <param name="UseOpenGL"            type="bool" value="true" />
    <!-- CUDA parameter -->
    <param name="EnableCUDA"            type="bool" value="true" />
  </node>
</launch>
